### PR TITLE
✨[patch] Implementation of agent-sidecar storage backup logic

### PR DESCRIPTION
**Pull request by rinx**
### Description:

In this PR, a blob storage backup logic of agent-sidecar is implemented.
Currently, it only supports S3.

Bugfix:
- internal/core/ngt/ngt.go ... remove useless `C.free`s

### Related Issue:

Nothing.

### How Has This Been Tested?:

Nothing.

### Environment:

- Golang Version: 1.14.3
- Docker Version: 19.03.8
- Kubernetes Version: 1.18.2
- NGT Version: 1.11.5

### Types of changes:

- [ ] Bug fix [type/bug]
- [X] New feature [type/feature]
- [ ] Add tests [type/test]
- [ ] Security related changes [type/security]
- [ ] Add documents [type/documentation]
- [ ] Refactoring [type/refactoring]
- [ ] Update dependencies [type/dependency]
- [ ] Update benchmarks and performances [type/bench]
- [ ] Update CI [type/ci]

### Changes to Core Features:

- [ ] Have you added an explanation of what your changes do and why you'd like us to include them?
- [ ] Have you written new tests for your core changes, as applicable?
- [ ] Have you successfully ran tests with your changes locally?

### Checklist:

- [X] I have read the [CONTRIBUTING](https://github.com/vdaas/vald/blob/master/CONTRIBUTING.md) document.
- [X] I have checked open [Pull Requests](https://github.com/vdaas/vald/pulls) for the similar feature or fixes?
- [ ] I have added tests and benchmarks to cover my changes.
- [ ] I have ensured all new and existing tests passed.
- [ ] I have commented my code, particularly in hard-to-understand areas
- [ ] I have updated the documentation accordingly.

### CODE REVIEW

File: internal/core/ngt/ngt.go [hunk 1]
```diff
@@ -534,10 +534,10 @@ func (n *ngt) Close() {
...
 		// ospace              C.NGTObjectSpace
-		C.free(unsafe.Pointer(&n.dimension))
-		C.free(unsafe.Pointer(&n.prop))
-		C.free(unsafe.Pointer(&n.ebuf))
+		// C.free(unsafe.Pointer(&n.dimension))
+		// C.free(unsafe.Pointer(&n.prop))
+		// C.free(unsafe.Pointer(&n.ebuf))
 		// C.free(unsafe.Pointer(&n.index))
-		C.free(unsafe.Pointer(&n.ospace))
+		// C.free(unsafe.Pointer(&n.ospace))
```
Comment (rinx) [inline internal/core/ngt/ngt.go:541]: Just commented out.
Should I remove these lines?

File: pkg/agent/sidecar/service/storage_observer.go [hunk 2]
```diff
@@ -127,17 +175,66 @@ func (o *observer) Start(ctx context.Context) (<-chan error, error) {
...
+func (o *observer) backup(ctx context.Context) error {
 	ctx, span := trace.StartSpan(ctx, "vald/agent-sidecar/service/StorageObserver.backup")
 	defer func() {
 		if span != nil {
 			span.End()
 		}
 	}()
-	// TODO implement backup logic here
+
+	sw, err := o.storage.Writer(ctx)
```
Comment (rinx) [inline pkg/agent/sidecar/service/storage_observer.go:190]: badge
Use io.Pipe

File: pkg/agent/sidecar/service/storage_observer.go [hunk 3]
```diff
@@ -127,17 +175,66 @@ func (o *observer) Start(ctx context.Context) (<-chan error, error) {
...
+			if err != nil {
+				return err
+			}
+
+			if !fi.IsDir() {
+				data, err := os.Open(file)
+				if err != nil {
+					return err
+				}
+				_, err = io.Copy(tw, data)
```
Comment (rinx) [inline pkg/agent/sidecar/service/storage_observer.go:226]: badge
Alos use io.Pipe

File: internal/config/sidecar.go [hunk 4]
```diff
@@ -18,8 +18,6 @@
 package config
 
 type AgentSidecar struct {
-	// Name string `yaml:"name" json:"name"`
-
 	// WatchPaths represents watch path list for backup
 	WatchPaths []string `yaml:"watch_paths" json:"watch_paths"`
```
Comment (rinx) [inline internal/config/sidecar.go:22]: badge
It should be single path `string` to specify ngt-index dir

File: internal/db/storage/blob/s3/s3.go [hunk 5]
```diff
@@ -17,59 +17,63 @@
...
-
-type sess struct {
-	endpoint        string
-	region          string
-	accessKey       string
-	secretAccessKey string
-	token           string
-	useLegacyList   bool
-}
+type s3client struct {
```
Comment (kpango) [inline internal/db/storage/blob/s3/s3.go:31]: I think s3.client is better than s3.s3client
```suggestion
type client struct {
```

File: internal/db/storage/blob/s3/writer/writer.go [hunk 6]
```diff
@@ -0,0 +1,124 @@
...
+	"github.com/aws/aws-sdk-go/service/s3/s3manager"
+	"github.com/vdaas/vald/internal/errgroup"
+	"github.com/vdaas/vald/internal/errors"
+	"github.com/vdaas/vald/internal/log"
+	"github.com/vdaas/vald/internal/safety"
+)
+
+type writer struct {
+	eg      errgroup.Group
+	service *s3.S3
```
Comment (kpango) [inline internal/db/storage/blob/s3/writer/writer.go:35]: do we need this field?

File: internal/db/storage/blob/s3/s3.go (hunk 5 above)
Comment (rinx) [reply internal/db/storage/blob/s3/s3.go:31]: Thanks. I'll rename it.

File: internal/db/storage/blob/s3/writer/writer.go (hunk 6 above)
Comment (rinx) [reply internal/db/storage/blob/s3/writer/writer.go:35]: we need it at L104 to initialize uploader.
https://github.com/vdaas/vald/pull/409/files#diff-da7d1c78589548bbb90ba2560947237bR104

**Pull request merged**